rust/rebase/src/flow/poap_ownership.rs

```diff
@@ -0,0 +1,193 @@
+use crate::{
+    content::poap_ownership::PoapOwnership as Ctnt,
+    proof::poap_ownership::PoapOwnership as Prf,
+    statement::poap_ownership::PoapOwnership as Stmt,
+    types::{
+        enums::subject::{Pkh, Subjects},
+        error::FlowError,
+        types::{Flow, FlowResponse, Instructions, Issuer, Proof, Statement, Subject},
+    },
+};
+
+use async_trait::async_trait;
+use chrono::{DateTime, Duration, Utc};
+use reqwest::{
+    header::{HeaderMap, HeaderName, HeaderValue},
+    Client,
+};
+use schemars::schema_for;
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+#[derive(Clone, Deserialize, Serialize)]
+pub struct PoapOwnership {
+    api_key: String,
+    // The amount of time that can pass before the witness
+    // wants a new flow initiated. In demo, set to 15 mins.
+    // This is checked for a negative value or 0 and errs if one is found
+    // Alternative is casting u64 to i64 and risking UB.
+    max_elapsed_minutes: i64,
+}
+
+impl PoapOwnership {
+    // NOTE: This method would be vulnerable to someone foward-dating signatures.
+    // It likely wouldn't occur, but could be mitigated by doing a Challenge { challenge: string, timestamp: string}
+    // and attaching it to FlowResponse with a challenge: Option<Challenge> field.
+    // Then, we generate the TS here, like in email, but send it back over the wire as part of the statement.
+    // That said, there's no motivation for commiting that style of attack in the current NFT gating demo situation.
+    // NOTE: People with clocks that are off might mess this up too.
+    // TODO: When moving to post-demo impl, rework this to use the above strategy
+    pub fn sanity_check(&self, timestamp: &str) -> Result<(), FlowError> {
+        if self.max_elapsed_minutes <= 0 {
+            return Err(FlowError::Validation(
+                "Max elapsed minutes must be set to a number greater than 0".to_string(),
+            ));
+        }
+
+        let now = Utc::now();
+        let then = DateTime::parse_from_rfc3339(timestamp)
+            .map_err(|e| FlowError::Validation(e.to_string()))?;
+
+        if then > now {
+            return Err(FlowError::Validation(format!(
+                "Timestamp provided comes from the future"
+            )));
+        }
+
+        if now - Duration::minutes(self.max_elapsed_minutes) > then {
+            return Err(FlowError::Validation(format!(
+                "Validation window has expired"
+            )));
+        };
+        Ok(())
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+struct PoapResEntry {
+    event: PoapEventEntry,
+    #[serde(rename = "tokenId")]
+    token_id: String,
+    owner: String,
+    chain: String,
+    // NOTE: This date is in the format "YYYY-MM-DD HH-mm-ss"
+    created: String,
```

Comment:
I don't think this is used. You don't have to define the whole struct, serde will still deserialise it